internal/provider/cached_image_resource.go

```diff
@@ -293,128 +295,246 @@ func (r *CachedImageResource) Configure(ctx context.Context, req resource.Config
-	// Other values set by the provider may be overridden, but will generate a warning.
-	var diag, ds diag.Diagnostics
-	if !data.ExtraEnv.IsNull() {
-		for key, elem := range data.ExtraEnv.Elements() {
-			switch key {
-			// These are required and should not be overridden.
-			case "ENVBUILDER_CACHE_REPO", "ENVBUILDER_GIT_URL":
-				diag.AddAttributeWarning(path.Root("extra_env"),
-					"Cannot override required environment variable",
-					fmt.Sprintf("The key %q in extra_env cannot be overridden.", key),
+	// convert extraEnv to a map for ease of use.
+	extraEnv := make(map[string]string)
+	for k, v := range data.ExtraEnv.Elements() {
+		extraEnv[k] = tfValueToString(v)
+	}
+	diags = append(diags, overrideOptionsFromExtraEnv(&opts, extraEnv, overrides)...)
+
+	return opts, diags
+}
+
+func overrideOptionsFromExtraEnv(opts *eboptions.Options, extraEnv map[string]string, overrides map[string]struct{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	// Make a map of the options for easy lookup.
+	optsMap := make(map[string]pflag.Value)
+	for _, opt := range opts.CLI() {
+		optsMap[opt.Env] = opt.Value
+	}
+	for key, val := range extraEnv {
+		switch key {
+
+		// These options may not be overridden.
+		case "ENVBUILDER_CACHE_REPO", "ENVBUILDER_GIT_URL":
+			diags.AddAttributeWarning(path.Root("extra_env"),
+				"Cannot override required environment variable",
+				fmt.Sprintf("The key %q in extra_env cannot be overridden.", key),
+			)
+			continue
+
+		default:
+			// Check if the option was set on the provider data model and generate a warning if so.
+			if _, overridden := overrides[key]; overridden {
+				diags.AddAttributeWarning(path.Root("extra_env"),
+					"Overriding provider environment variable",
+					fmt.Sprintf("The key %q in extra_env overrides an option set on the provider.", key),
+				)
+			}
+
+			// XXX: workaround for serpent behaviour where calling Set() on a
+			// string slice will append instead of replace: set to empty first.
+			if key == "ENVBUILDER_IGNORE_PATHS" {
+				_ = optsMap[key].Set("")
+			}
+
+			opt, found := optsMap[key]
+			if !found {
+				// ignore unknown keys
+				continue
+			}
+
+			if err := opt.Set(val); err != nil {
+				diags.AddAttributeError(path.Root("extra_env"),
+					"Invalid value for environment variable",
+					fmt.Sprintf("The key %q in extra_env has an invalid value: %s", key, err),
 				)
-			default:
-				if _, ok := env[key]; ok {
-					// This is a warning because it's possible that the user wants to override
-					// a value set by the provider.
-					diag.AddAttributeWarning(path.Root("extra_env"),
-						"Overriding provider environment variable",
-						fmt.Sprintf("The key %q in extra_env overrides an environment variable set by the provider.", key),
-					)
-				}
-				env[key] = tfValueToString(elem)
 			}
 		}
 	}
+	return diags
+}
+
+func computeEnvFromOptions(opts eboptions.Options, extraEnv map[string]string) map[string]string {
```

Comment:
nit: Maybe we should move `computeEnvFromOptions` to a helper function or lib as `cached_image_resource.go` is getting larger.